tests/php/AssetControlExtensionTest.php

```diff
@@ -222,6 +224,95 @@ public function testReplaceFile()
         $this->assertEquals(AssetStore::VISIBILITY_PUBLIC, $object3->Header->getVisibility());
     }
 
+    
+    public function testReplaceWithVariant()
+    {
+        $store = $this->getAssetStore();
+        $object1 = AssetControlExtensionTest\VersionedObject::get()->filter('Title', 'My object')->first();
+        /** @var DBFile $download */
+        $download = $object1->Download;
+        Versioned::set_stage(Versioned::DRAFT);
+        $v2Content = 'Documents/File.txt v2';
+        $v3Content = 'Documents/File.txt v3';
+
+        $v1 = new ParsedFileID($download->getFilename(), $download->getHash());
+        $v2 = $v1->setHash(sha1($v2Content));
+        $v3 = $v1->setHash(sha1($v3Content));
+
+        // Start by creating a pre-existing variant
+        $download->setFromString(
+            'Variant of Documents/File.txt',
+            $v1->getFilename(),
+            $v1->getHash(),
+            'boom'
+        );
+        $this->assertTrue(
+            $store->exists($v1->getFilename(), $v1->getHash(), 'boom'),
+            sprintf('A variant of %s has been created', $v1->getFilename())
+        );
+        
+        // Let's replace the content of the main file and publish it
+        $download->setFromString($v2Content, $v2->getFilename());
+        $object1->write();
```

Comment:
I tried not having the `write()` there and just use the publish, but that created some weird results.

I'll investigate this a bit further.

Reply:
Everywhere, we play with files, we always write them first before publishing them. It seems this has been a constant limitation we've been working around.

It probably would be more intuitive to allow people to publish right away, but that's probably a different bug for another day.

I'll raise an issue for it.